[BUG] org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testDropPrimaryDuringReplication is flaky

**Describe the bug**
`org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testDropPrimaryDuringReplication` is flaky

https://build.ci.opensearch.org/job/gradle-check/17543/testReport/junit/org.opensearch.remotestore/SegmentReplicationUsingRemoteStoreIT/testDropPrimaryDuringReplication/

https://github.com/opensearch-project/OpenSearch/pull/8057#issuecomment-1590873556

**Assertion Failure**
```
java.lang.AssertionError: Expected search hits on node: node_t3 to be at least 111 but was: 110
	at __randomizedtesting.SeedInfo.seed([5E122859418B05B6:F3568B11F61BBC1C]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.opensearch.indices.replication.SegmentReplicationBaseIT.lambda$waitForSearchableDocs$0(SegmentReplicationBaseIT.java:122)
	at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1084)
	at org.opensearch.indices.replication.SegmentReplicationBaseIT.waitForSearchableDocs(SegmentReplicationBaseIT.java:117)
	at org.opensearch.indices.replication.SegmentReplicationBaseIT.waitForSearchableDocs(SegmentReplicationBaseIT.java:112)
	at org.opensearch.indices.replication.SegmentReplicationIT.testDropPrimaryDuringReplication(SegmentReplicationIT.java:739)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
```

**To Reproduce**
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testDropPrimaryDuringReplication" -Dtests.seed=5E122859418B05B6 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-CO -Dtests.timezone=Asia/Bishkek -Druntime.java=20
```

--- a/server/src/main/java/org/opensearch/index/engine/NRTReplicationReaderManager.java
+++ b/server/src/main/java/org/opensearch/index/engine/NRTReplicationReaderManager.java
@@ -98,12 +98,12 @@ public class NRTReplicationReaderManager extends OpenSearchReaderManager {
      * @param infos {@link SegmentInfos} infos
      * @throws IOException - When Refresh fails with an IOException.
      */
-    public synchronized void updateSegments(SegmentInfos infos) throws IOException {
+    public void updateSegments(SegmentInfos infos) throws IOException {
         // roll over the currentInfo's generation, this ensures the on-disk gen
         // is always increased.
         infos.updateGeneration(currentInfos);
         currentInfos = infos;
-        maybeRefresh();
+        maybeRefreshBlocking();
     }
 
     public SegmentInfos getSegmentInfos() {

--- a/server/src/test/java/org/opensearch/index/engine/NRTReplicationReaderManagerTests.java
+++ b/server/src/test/java/org/opensearch/index/engine/NRTReplicationReaderManagerTests.java
@@ -1,0 +1,88 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.engine;
+
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.index.StandardDirectoryReader;
+import org.apache.lucene.util.Version;
+import org.opensearch.common.lucene.index.OpenSearchDirectoryReader;
+import org.opensearch.index.store.Store;
+
+import java.io.IOException;
+
+public class NRTReplicationReaderManagerTests extends EngineTestCase {
+
+    public void testCreateNRTreaderManager() throws IOException {
+        try (final Store store = createStore()) {
+            store.createEmpty(Version.LATEST);
+            final DirectoryReader reader = DirectoryReader.open(store.directory());
+            final SegmentInfos initialInfos = ((StandardDirectoryReader) reader).getSegmentInfos();
+            NRTReplicationReaderManager readerManager = new NRTReplicationReaderManager(
+                OpenSearchDirectoryReader.wrap(reader, shardId),
+                (files) -> {},
+                (files) -> {}
+            );
+            assertEquals(initialInfos, readerManager.getSegmentInfos());
+            try (final OpenSearchDirectoryReader acquire = readerManager.acquire()) {
+                assertNull(readerManager.refreshIfNeeded(acquire));
+            }
+
+            // create an updated infos
+            final SegmentInfos infos_2 = readerManager.getSegmentInfos().clone();
+            infos_2.changed();
+
+            readerManager.updateSegments(infos_2);
+            assertEquals(infos_2, readerManager.getSegmentInfos());
+            try (final OpenSearchDirectoryReader acquire = readerManager.acquire()) {
+                final StandardDirectoryReader standardReader = NRTReplicationReaderManager.unwrapStandardReader(acquire);
+                assertEquals(infos_2, standardReader.getSegmentInfos());
+            }
+        }
+    }
+
+    public void testUpdateSegmentsWhileRefreshing() throws IOException, InterruptedException {
+        try (final Store store = createStore()) {
+            store.createEmpty(Version.LATEST);
+            final DirectoryReader reader = DirectoryReader.open(store.directory());
+            NRTReplicationReaderManager readerManager = new NRTReplicationReaderManager(
+                OpenSearchDirectoryReader.wrap(reader, shardId),
+                (files) -> {},
+                (files) -> {}
+            );
+
+            final SegmentInfos infos_2 = readerManager.getSegmentInfos().clone();
+            infos_2.changed();
+
+            Thread refreshThread = new Thread(() -> {
+                try {
+                    readerManager.maybeRefresh();
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+            Thread updateThread = new Thread(() -> {
+                try {
+                    readerManager.updateSegments(infos_2);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+            refreshThread.start();
+            updateThread.start();
+            refreshThread.join();
+            updateThread.join();
+            try (final OpenSearchDirectoryReader acquire = readerManager.acquire()) {
+                final StandardDirectoryReader standardReader = NRTReplicationReaderManager.unwrapStandardReader(acquire);
+                assertEquals(infos_2.version, standardReader.getSegmentInfos().version);
+            }
+            assertEquals(infos_2, readerManager.getSegmentInfos());
+        }
+    }
+}

